### PR TITLE
a draft for the new update version function

### PR DESCRIPTION
**Pull request by viniciusdc**
As already mentioned in regro/cf-scripts#842, and cited by regro/cf-scripts#842 (comment), follows the firts attempt for the new update versions function.
-- i need to change the dir where `new_version.json` will be saved.
-- Someone run this, all my attempts were interrupted by some string error (in the end).

### CODE REVIEW

**Comment by CJ-Wright** [comment]
Please run [`black`](https://black.readthedocs.io/en/stable/) on this code when you get a chance.

File: conda_forge_tick/migrate_update_version.py [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+to_update["nodes"] = []
+for node in gx.nodes:
+    with gx.nodes["{}".format(node)]["payload"] as attrs:
+        # verify the actual situation of the package;
+        if attrs.get("bad") or attrs.get("archived"):
+            new_version_flag = False
+        
+        # new version request
+        try:
+            new_version = get_latest_version(node, attrs, sources)
```
Comment (CJ-Wright) [inline conda_forge_tick/migrate_update_version.py]: I wouldn't try to get the latest version if the feedstock has been marked bad/archived

File: conda_forge_tick/migrate_update_version.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+            new_version = new_version or new_version_flag
+        except Exception as e:
+            try:
+                se = repr(e)
+            except Exception as ee:
+                se = "Bad exception string: {}".format(ee)
+                print("warning: Error getting uptream version of {}: {}".format(node, se))
+        else:
+            print("# {:<5} - {:<30} - ver: {:<10} - new ver: {}".format(Node_count, node, attrs.get('version'), str(new_version)))
```
Comment (CJ-Wright) [inline conda_forge_tick/migrate_update_version.py]: Please use f strings

File: conda_forge_tick/migrate_update_version.py [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+        
+        to_update["nodes"].append({"id": str(node), "version": str(new_version)})
+    Node_count += 1
```
Comment (CJ-Wright) [inline conda_forge_tick/migrate_update_version.py]: Where is `Node_count` being used?

File: conda_forge_tick/migrate_update_version.py (hunk 1 above)
Comment (viniciusdc) [reply conda_forge_tick/migrate_update_version.py]: ok, i will put a flag for that, thanks.

File: conda_forge_tick/migrate_update_version.py (hunk 2 above)
Comment (viniciusdc) [reply conda_forge_tick/migrate_update_version.py]: yes, sorry about that

File: conda_forge_tick/migrate_update_version.py (hunk 3 above)
Comment (CJ-Wright) [reply conda_forge_tick/migrate_update_version.py]: Never mind

Comment (viniciusdc) [reply conda_forge_tick/migrate_update_version.py]: inside the print in `else`, just to see the progress, i can take it off if you want.

File: conda_forge_tick/migrate_update_version.py (hunk 1 above)
Comment (viniciusdc) [reply conda_forge_tick/migrate_update_version.py]: Done,

File: conda_forge_tick/migrate_update_version.py [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+_all_nodes = [t for t in gx.nodes.items()]
+random.shuffle(_all_nodes)
+
+# from here you can inspect the graph object, and what node will be updated:
+print(f"Number of nodes: {len(gx.nodes)}")
+Node_count = 0
+
+to_update = {}
+to_update["nodes"] = []
+for node, node_attrs in _all_nodes:
```
Comment (viniciusdc) [inline conda_forge_tick/migrate_update_version.py]: Now i am using the `_all_nodes`

Comment (viniciusdc) [reply conda_forge_tick/migrate_update_version.py]: I still need to specify an dir to write the .json file. (But i don't know where)

File: conda_forge_tick/migrate_update_version.py [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+    with node_attrs["payload"] as attrs:
+        # verify the actual situation of the package;
+        actual_ver = str(attrs.get("version"))
+        if attrs.get("bad") or attrs.get("archived"):
+            print(
+                f"# {Node_count:<5} - {node:<30} - ver: {actual_ver:<10} - bad/archived"
+            )
+            pass
```
Comment (CJ-Wright) [inline conda_forge_tick/migrate_update_version.py]: you need `continue` here not pass

Comment (viniciusdc) [reply conda_forge_tick/migrate_update_version.py]: alright i will change that, thanks

**Comment by beckermr** [comment]
We need to think about the output format a bit. Do we want "id"? Maybe "node"? Also, we should make sure this data is dynamodb friendly etc.

**Comment by viniciusdc** [comment]
> We need to think about the output format a bit. Do we want "id"? Maybe "node"? Also, we should make sure this data is dynamodb friendly etc.

Yes, i was using the `{"node":{"id":new version}}` structure for the json file, as i saw something similar in the graph.json, sorry about that. For the Dynamodb part i still dont know exactly what it is.

**Comment by viniciusdc** [comment]
Just to give a feedback, what I am doing now, looking for more information on the Dynamodb tables and doing some experiments with it, maybe tomorrow i will have something ...
I am using this site as source: https://www.dynamodbguide.com

**Comment by viniciusdc** [comment]
I've already created a local Dynamodb and I will try this structure for the .json```
{
	"Nodes": 
		"M": 	{
			"node_id": {"version": "3.1"},
			 ...}
			}
}
```

**Comment by viniciusdc** [comment]
Hello, sorry for the late feedback (I needed to make some stuff because the covid). Just a simple doubt, we are going to export just the data into Dynamodb tables, or it will accept scripts as well?

**Comment by viniciusdc** [comment]
And I am following the json format as specified here: https://docs.aws.amazon.com/cli/latest/reference/dynamodb/put-item.html

**Comment by beckermr** [comment]
No worries! Stay healthy! CC @scopatz on the dynamodb format stuff since he is more familiar with it.

**Comment by viniciusdc** [comment]
Tks, I will do it. Any other task i can do ? "Stay healthy!" yes thanks, you too 😄

**Comment by viniciusdc** [comment]
can I use this as a model ?
https://github.com/regro/cf-scripts/blob/master/conda_forge_tick/dynamo_models.py

**Comment by CJ-Wright** [comment]
you'll need to modify it some, but you can use it as a start

**Comment by viniciusdc** [comment]
> you'll need to modify it some, but you can use it as a start

Thanks, i will do 👍

**Comment by CJ-Wright** [comment]
Do want to handle the dynamo parts as an independent PR? I don't want to slow down this PR if possible.

**Comment by viniciusdc** [comment]
Ok, we can do it like you said. I just thought that the next part of my work was to set up the dynamodb format of the json output.

**Comment by CJ-Wright** [comment]
Absolutely, but each PR should ideally implement one feature so you don't get bogged down.

**Comment by viniciusdc** [comment]
I am using only a simple scheme for the JSON output file, as we are giving a time with the DynamoDB stuff.

**Comment by viniciusdc** [comment]
Hello, I get this error when testing, I understand what it means but I want to know if is there something I can do to fix (probably something is still missing when I separated the sources) ?
```python
WARNING  conda_forge_tick._update_versions || Warning: Error getting upstream version of vigra: AttributeError("'NoneType' object has no attribute 'group'")

```

**Comment by CJ-Wright** [comment]
That looks like a regex gone wrong from here: https://github.com/regro/cf-scripts/pull/907/files#diff-48e946714461e4ec5ef529b2442783a4R55

**Comment by viniciusdc** [comment]
Thank you !!! I will look into that, as soon I finish some tests here. Ps. Sorry for the last meeting, I always become nervous...kkkk

**Comment by CJ-Wright** [comment]
Also it is possible that the recipe is not compatible with the bot. I would check this against the bot's current performance.

**Comment by viniciusdc** [comment]
> Also it is possible that the recipe is not compatible with the bot. I would check this against the bot's current performance.

IDK, it happened with another pck too, I am trying to setup a separted environment to test this behaviorr.

**Comment by viniciusdc** [comment]
I tested the code, using a copy repo from  `cf-scripts` and the graph from `cf-county-fair` and it worked  !!! Here's a link for the Circle CI test: https://app.circleci.com/pipelines/github/viniciusdc/ts-graph/46/workflows/cf6da52d-df5d-485d-9aaf-6c142ce10962/jobs/51

**Comment by viniciusdc** [comment]
> I tested the code, using a copy repo from `cf-scripts` and the graph from `cf-county-fair` and it worked !!! Here's a link for the Circle CI test: https://app.circleci.com/pipelines/github/viniciusdc/ts-graph/46/workflows/cf6da52d-df5d-485d-9aaf-6c142ce10962/jobs/51

Circle CI, reaaly is an awsome tool !! But I have a doubt, I saw that organizations have 400.000 credits per month, is it something I can have too as a new user ? (I ask this cause when I acess my information from 'viniciusdc' user it say org information)

**Comment by viniciusdc** [comment]
Can I close this,  and open a new one with the last updates (just tested it on Circle and everything looks good) ?

**Comment by beckermr** [comment]
sure!

**Comment by CJ-Wright** [comment]
Just to lay the ground work for the next PR, here is a list of things that I think are needed (most of which I think you had in this one):
1. an addition to the `cli.xsh` file for the update version script
2. The update version script itself
3. The upstream sources as an independent module in the library
4. Changes to make graph (or maybe the version bump code) to look at the source of new version numbers. (I think this is the main piece missing from this PR)

**Comment by viniciusdc** [comment]
> Just to lay the ground work for the next PR, here is a list of things that I think are needed (most of which I think you had in this one):
> 
> 1. an addition to the `cli.xsh` file for the update version script
> 2. The update version script itself
> 3. The upstream sources as an independent module in the library
> 4. Changes to make graph (or maybe the version bump code) to look at the source of new version numbers. (I think this is the main piece missing from this PR)

Oh thank you very much. The 2 and 3 items are done, The 4 I didn't understand exactly, but I will make the new PR and work forward.

**Comment by CJ-Wright** [comment]
My understanding of the current plan is to have an independent file, created by and independent worker process, to house the latest version numbers of all the source code we package. This is a departure from how that information is currently stored (in the node_attrs). That means that pieces of the code that need the new version numbers (for instance the version bump code) will need to be rewritten to look at the new file.

**Comment by viniciusdc** [comment]
Ahh, I've not seen that function yet, thank you I will look into that for the new update.

**Pull request closed without merging**